**Assignment Number:** HW5

**Name:** Nghia Lam

**UTA ID:** 1001699317


## INSTRUCTIONS 

Every learner should submit his/her own homework solutions. However, you are allowed to discuss the homework with each other– but everyone must submit his/her own solution; you may not copy someone else’s solution. 

The homework consists of two parts:
1.	Data manipulation
2.	Multipe regression Analysis

Follow the prompts in the attached jupyter notebook. Download the data and place it in your working directory, or modify the path to upload it to your notebook. Add markdown cells to your analysis to include your solutions, comments, answers. **Add as many cells as you need**, for easy readability comment when possible.
Hopefully this homework will help you develop skills, make you understand the flow of an EDA, get you ready for individual work.

Submission: Run all your code cells and export the file as HTML. Submit a zip of your .ipynb file and HTML file. 

Good luck!


**Title: 1985 Auto Imports Database**

Relevant Information:
   -- Description
      This data set consists of three types of entities: (a) the
      specification of an auto in terms of various characteristics, (b)
      its assigned insurance risk rating, (c) its normalized losses in use
      as compared to other cars.  The second rating corresponds to the
      degree to which the auto is more risky than its price indicates.
      Cars are initially assigned a risk factor symbol associated with its
      price.   Then, if it is more risky (or less), this symbol is
      adjusted by moving it up (or down) the scale.  Actuarians call this
      process "symboling".  A value of +3 indicates that the auto is
      risky, -3 that it is probably pretty safe.

      The third factor is the relative average loss payment per insured
      vehicle year.  This value is normalized for all autos within a
      particular size classification (two-door small, station wagons,
      sports/speciality, etc...), and represents the average loss per car
      per year.

   -- Note: Several of the attributes in the database could be used as a
            "class" attribute.

5. Number of Instances: 205

6. Number of Attributes: 26 total
   -- 15 continuous
   -- 1 integer
   -- 10 nominal

7. Attribute Information:     
     Attribute:                Attribute Range:
     ------------------        -----------------------------------------------
  1. symboling:                -3, -2, -1, 0, 1, 2, 3.
  2. normalized-losses:        continuous from 65 to 256.
  3. make:                     alfa-romero, audi, bmw, chevrolet, dodge, honda,isuzu, jaguar, mazda, mercedes-benz, mercury, mitsubishi, nissan, peugot, plymouth, porsche, renault, saab, subaru, toyota,volkswagen, volvo
  4. fuel-type:                diesel, gas.
  5. aspiration:               std, turbo.
  6. num-of-doors:             four, two.
  7. body-style:               hardtop, wagon, sedan, hatchback, convertible.
  8. drive-wheels:             4wd, fwd, rwd.
  9. engine-location:          front, rear.
 10. wheel-base:               continuous from 86.6 120.9.
 11. length:                   continuous from 141.1 to 208.1.
 12. width:                    continuous from 60.3 to 72.3.
 13. height:                   continuous from 47.8 to 59.8.
 14. curb-weight:              continuous from 1488 to 4066.
 15. engine-type:              dohc, dohcv, l, ohc, ohcf, ohcv, rotor.
 16. num-of-cylinders:         eight, five, four, six, three, twelve, two.
 17. engine-size:              continuous from 61 to 326.
 18. fuel-system:              1bbl, 2bbl, 4bbl, idi, mfi, mpfi, spdi, spfi.
 19. bore:                     continuous from 2.54 to 3.94.
 20. stroke:                   continuous from 2.07 to 4.17.
 21. compression-ratio:        continuous from 7 to 23.
 22. horsepower:               continuous from 48 to 288.
 23. peak-rpm:                 continuous from 4150 to 6600.
 24. city-mpg:                 continuous from 13 to 49.
 25. highway-mpg:              continuous from 16 to 54.
 26. price:                    continuous from 5118 to 45400.

8. Missing Attribute Values: (denoted by "?")

In [44]:
from scipy import stats
from sklearn.linear_model import LinearRegression
from statsmodels.compat import lzip
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm
from statsmodels.stats.outliers_influence import variance_inflation_factor
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm

%matplotlib inline

In [45]:
#Read in data
df2 =pd.read_csv('df2.csv')

df2.head()

fuel_type  wheel_base  length  width  heights  curb_weight  engine_size  \
0       gas        88.6   168.8   64.1     48.8         2548          130   
1       gas        88.6   168.8   64.1     48.8         2548          130   
2       gas        94.5   171.2   65.5     52.4         2823          152   
3       gas        99.8   176.6   66.2     54.3         2337          109   
4       gas        99.4   176.6   66.4     54.3         2824          136   

   bore  stroke  comprassion  horse_power  peak_rpm  city_mpg  highway_mpg  \
0  3.47    2.68          9.0          111      5000        21           27   
1  3.47    2.68          9.0          111      5000        21           27   
2  2.68    3.47          9.0          154      5000        19           26   
3  3.19    3.40         10.0          102      5500        24           30   
4  3.19    3.40          8.0          115      5500        18           22   

   price  
0  13495  
1  16500  
2  16500  
3  13950  
4  17450

## 1. Data Manipulation

In [46]:
df2.shape

(195, 15)

### Get dummy variables for fuel_type

In [47]:
##Use get_dummies() and drop_first=True
##Your code goes here
df2 = pd.get_dummies(df2, columns=['fuel_type'], drop_first=True)

In [48]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   wheel_base     195 non-null    float64
 1   length         195 non-null    float64
 2   width          195 non-null    float64
 3   heights        195 non-null    float64
 4   curb_weight    195 non-null    int64  
 5   engine_size    195 non-null    int64  
 6   bore           195 non-null    float64
 7   stroke         195 non-null    float64
 8   comprassion    195 non-null    float64
 9   horse_power    195 non-null    int64  
 10  peak_rpm       195 non-null    int64  
 11  city_mpg       195 non-null    int64  
 12  highway_mpg    195 non-null    int64  
 13  price          195 non-null    int64  
 14  fuel_type_gas  195 non-null    uint8  
dtypes: float64(7), int64(7), uint8(1)
memory usage: 21.6 KB


In [49]:
df2.head()

wheel_base  length  width  heights  curb_weight  engine_size  bore  stroke  \
0        88.6   168.8   64.1     48.8         2548          130  3.47    2.68   
1        88.6   168.8   64.1     48.8         2548          130  3.47    2.68   
2        94.5   171.2   65.5     52.4         2823          152  2.68    3.47   
3        99.8   176.6   66.2     54.3         2337          109  3.19    3.40   
4        99.4   176.6   66.4     54.3         2824          136  3.19    3.40   

   comprassion  horse_power  peak_rpm  city_mpg  highway_mpg  price  \
0          9.0          111      5000        21           27  13495   
1          9.0          111      5000        21           27  16500   
2          9.0          154      5000        19           26  16500   
3         10.0          102      5500        24           30  13950   
4          8.0          115      5500        18           22  17450   

   fuel_type_gas  
0              1  
1              1  
2              1  
3              1  
4              1

## 2. Multiple Regression Analysis

**1. Create a model that uses all the variables** and call it model1. The dependent variable is price, the independent variables are all the rest. Print out a summary of the model (coefficents, stanrard errors, confidence intervals and other metrics shown in class and answer the quesions based on your output.

In [50]:
## your code goes here

#used df3 because fuel was an object in df2 so I used the dummy data intead
X = df2.drop('price',axis=1)
y = df2['price']

#Add a constant term for the intercept (β0)
X = sm.add_constant(X)
reg = LinearRegression()
reg.fit(X,y)
#print(reg.intercept_)
#print(df3.dtypes)

model1 = sm.OLS(y,X).fit()

print(f"\nredidual std error: {model1.mse_resid ** 0.5}\n")
print(model1.summary())





redidual std error: 3130.8999955911945

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.860
Model:                            OLS   Adj. R-squared:                  0.849
Method:                 Least Squares   F-statistic:                     78.89
Date:                Wed, 27 Sep 2023   Prob (F-statistic):           5.84e-69
Time:                        00:17:57   Log-Likelihood:                -1838.5
No. Observations:                 195   AIC:                             3707.
Df Residuals:                     180   BIC:                             3756.
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const

c:\Users\nghia\anaconda3\lib\site-packages\statsmodels\tsa\tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


1. How do you interpret the intercept?
- In this case the intercept is -4.45e+04 or -44,500. This is the estimated price of the car. 
- This however does not make sense 

2. How many variables are statistically significant?
- There are 9 signifcant values 
    3 less then .05
    3 less then 0

3. What is the variance of the model?
- (Residual standard error: 3130.9)^2= 9,802,534.81

4. What is the coefficeint of determination and how do you interpret it?
- coefficient of determination is our R-squred. 
- our multiple sqare error is .860 which means that these independent values determine 86% of the price of the car

5. What is the F-statistics used for? How do you interpret it for this model?
- It is used to assess the overall significance of a regression model
- F-statistic: 78.89

**2. Dropp all the variables that are not statistically significant** at least at 90% confidence level. Run another regression model with price as the dependent variable and the rest of the variables as the independent variabls. Call it model2. Print a summary of the results and answer the questions bellow. 

In [51]:
## your code goes here
sigVar = model1.pvalues[model1.pvalues <= 0.10].index
Xsig = X[sigVar]

model2 = sm.OLS(y,Xsig).fit()
print(f"\nredidual std error: {model2.mse_resid ** 0.5}\n")
print(model2.summary())


redidual std error: 3138.8781611491454

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.854
Model:                            OLS   Adj. R-squared:                  0.848
Method:                 Least Squares   F-statistic:                     136.5
Date:                Wed, 27 Sep 2023   Prob (F-statistic):           1.29e-73
Time:                        00:17:57   Log-Likelihood:                -1842.2
No. Observations:                 195   AIC:                             3702.
Df Residuals:                     186   BIC:                             3732.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const

1. How do you interpret the intercept? 
- -6.156e+04 : This is now the new estimated price of the car at the intercept

2. How many variables are statistically significant? 
- There are 7 variables that are statiscticlly significant and different intervals (not including intercept)

3. What is the variance of the model? 
- (redidual std error: 3138.8781611491454)^2 = 9852555.72666

4. What is the coefficeint of determination and how do you interpret it? What is the Adjusted R-squared and compare it to the model1's value. 
- The coeffiiceint of determination is our R^2
- It is around 85% which means these independent values determine 85% of our car
- With these 8 independent values we are able to get 85%, this is with model 2. 
- With model1 we get around 86% with 14 independent values. 
- That means that there were alot of non significant values in our dataset. and from 8 to 14 values only increase by 1%

5. What is the F-statistics used for? How do you interpret it for this model?
- It is used to assess the overall significance of a regression model
- F-statistic: 136.5, this means that the whole model is very significant 

**3. Compare the two models with ANOVA**.

In [52]:
##your code goes here

anova_results = anova_lm(model2, model1)
print(anova_results)


   df_resid           ssr  df_diff       ss_diff        F  Pr(>F)
0     186.0  1.832575e+09      0.0           NaN      NaN     NaN
1     180.0  1.764456e+09      6.0  6.811918e+07  1.15819  0.3308


1. What are your null and alternative hypothesis? 
- Ho: Bi = 0 vs Ha: Bi != 0
- Ho means that there is no significant difference in model fit between the 2 models
- Ha means that there is a significant difference in model fit between the 2, thus one of them is better than the other
2. What is your conclusion?
- based on the P values (.3308) is greater than .05, you would fail to reject the null hypothesis
- it also looks like both intercept is also significant in this as p values are small and t values are large

### Answer